### PR TITLE
CDEPS update

### PR DESCRIPTION
**Pull request by maddenp-cu**
Updates to align the CDEPS driver with current `uwtools`. Adds tests and documentation.

### CODE REVIEW

File: docs/conf.py
Comment (maddenp-cu) [inline docs/conf.py]: This is a base URL that's used in the HTML documentation (the `.rst` files shown below, which Sphinx converts into HTML). If there's a better set of documentation for CDEPS, please let me know and I can update. There are also some question, below, about specific places where better links to documentation are probably needed.

File: docs/sections/user_guide/cli/drivers/cdeps.rst
Comment (maddenp-cu) [inline docs/sections/user_guide/cli/drivers/cdeps.rst:4]: This will link to https://escomp.github.io/CDEPS/versions/master/html/index.html for high-level overview documentation.

Also, when we merge this PR onto https://github.com/ufs-community/uwtools/pull/522, Read the Docs will build and stage a preview of the documentation that will be easier to review than this. I'll point out where to find those docs in the master PR after this merges.

File: docs/sections/user_guide/yaml/components/cdeps.rst
Comment (maddenp-cu) [inline docs/sections/user_guide/yaml/components/cdeps.rst]: The `???  HERE ???` marker needs to be replaced. Is there online documentation that describes the namelist options and their meanings? For example, in the [FV3 UW YAML docs](https://uwtools.readthedocs.io/en/main/sections/user_guide/yaml/components/fv3.html), in the `namelist:` section, is a link to [these docs](https://ufs-weather-model.readthedocs.io/en/latest/InputsOutputs.html#namelist-file-input-nml) outlining FV3's `input.nml` contents. Is there something similar for CDEPS? Otherwise, we may need a follow-up PR to add descriptions for the namelist options as specified in the UW YAML config.

If we can link to existing external documentation, I can make those changes in this PR.

File: docs/sections/user_guide/yaml/components/cdeps.rst
Comment (maddenp-cu) [inline docs/sections/user_guide/yaml/components/cdeps.rst]: Here, too, we should add descriptions of the stream-related UW YAML items. Since these are currently variables for use in a Jinja2 template, maybe there's no appropriate external documentation, in which case we'll just need to list and describe them here (but possibly back in the main PR targeting the `ufs-community` `uwtools` repo).  But if there's external documentation we can reference, we could possibly get away with just a table mapping our UW YAML names (`dtlimit`, `mapalgo`, etc.) to the official docs. Or, if there's a one-to-one mapping from UW YAML names to documented names, that's even better.

Essentially, though, we need to document what variables need to be set in YAML in e.g. `cdeps.atm_streams.streams0`.

File: docs/sections/user_guide/yaml/components/cdeps.rst
Comment (maddenp-cu) [inline docs/sections/user_guide/yaml/components/cdeps.rst]: Similarly, a better reference needed in place of `??? HERE ???`.

File: docs/shared/cdeps.yaml
Comment (maddenp-cu) [inline docs/shared/cdeps.yaml:2]: In combination with the updated `cdeps.jsonschema` (see below), this file now passes validation:
```
(DEV-uwtools) ~/git/uwtools $ uw config validate --schema-file src/uwtools/resources/jsonschema/cdeps.jsonschema --input-file docs/shared/cdeps.yaml 
[2024-07-17T16:06:16]     INFO 0 UW schema-validation errors found
```
I also ran it through a utility called `yaml-sort` to make it consistent with our other example files in `docs/shared`.

File: src/uwtools/cli.py
Comment (maddenp-cu) [inline src/uwtools/cli.py:855]: I made some changes in this module to support executing an assets-only driver (i.e. one that does not execute a component, but just provisions content to a run directory) like `cdeps` via the command line. Our other two existing assets-only drivers,`schism` and `ww3`, are slated for use only in a coupled driver and are not connected to the CLI, so this is the first time the need has arisen.

File: src/uwtools/drivers/cdeps.py
Comment (maddenp-cu) [inline src/uwtools/drivers/cdeps.py:16]: On `uwtools` `main`, we recently introduced several more-precise driver classes that provide correct `__init__()` and `_taskname()` methods, so those no longer need to be defined in this class.

File: src/uwtools/drivers/cdeps.py
Comment (maddenp-cu) [inline src/uwtools/drivers/cdeps.py:67]: Just moved down in the file...

File: src/uwtools/drivers/cdeps.py
Comment (maddenp-cu) [inline src/uwtools/drivers/cdeps.py:106]: This method is required by the `Assets` base class that this driver ultimately inherits from. It might not be useful for CDEPS. If it's too confusing, we could reconsider making this method mandatory,  Please let me know if you think that this method should _not_ exist and I can take care of it later after the main PR merges to `uwtools` `main`.

File: src/uwtools/resources/jsonschema/cdeps.jsonschema
Comment (maddenp-cu) [inline src/uwtools/resources/jsonschema/cdeps.jsonschema:2]: I reworked the schema somewhat, and hope that I've captured the requirements correctly. When evaluating it, I'd suggest:
1. Reading through it to make sure it seems correct.
2. Noting that it successfully validates, as mentioned in a previous comment, the `docs/shared/cdeps.yaml` example file.
3. Looking at the new unit tests in `test_schemas.py`, below, which demonstrates many things that pass and fail when validated with this schema, or with parts of it.

It may also help to just read the file as-is, without considering the diff against the previous version.

File: src/uwtools/resources/jsonschema/cdeps.jsonschema
Comment (maddenp-cu) [inline src/uwtools/resources/jsonschema/cdeps.jsonschema:3]: Since it appears that the `atm_streams` and `ocn_streams` YAML blocks have the same requirements, I factored this block out and then reference it with `$ref` tags in the appropriate sections below.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (maddenp-cu) [inline src/uwtools/tests/drivers/test_cdeps.py:3]: These tests are similar to those for other drivers, and execute the various task methods and helper methods in the module.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (maddenp-cu) [inline src/uwtools/tests/drivers/test_cdeps.py:94]: This test creates a simple Jinja2 template file that uses the values from the config to render to an output file, which is then checked against expected values.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (maddenp-cu) [inline src/uwtools/tests/drivers/test_cdeps.py:22]: The schema tests (see below) already define a CDEPS config for testing so, rather than define a new one here, it is imported from the `test_schemas` module.

File: src/uwtools/tests/test_schemas.py
Comment (maddenp-cu) [inline src/uwtools/tests/test_schemas.py:15]: This test config is used both for these schema tests, and for the driver tests in `test_cdeps.py`, above. The values used pass schema validation, but are mostly not very informative -- they do not represent realistic values, especially the ones that just say `"string"`. We could certainly update any of these with something more familiar, but I would need guidance on appropriate values.

File: src/uwtools/tests/test_schemas.py
Comment (maddenp-cu) [inline src/uwtools/tests/test_schemas.py:240]: The methods below test various parts of the CDEPS config schema in ways (and with comments) that hopefully illustrate the rules that the schema is trying to enforce. Any of this can be updated, now or later, if the schema needs to be corrected.

File: docs/conf.py
Comment (uturuncoglu) [reply docs/conf.py]: @maddenp-noaa Yes this ıs the maın documentation that I finalize recently as a part of an other project.

File: docs/sections/user_guide/cli/drivers/cdeps.rst
Comment (uturuncoglu) [reply docs/sections/user_guide/cli/drivers/cdeps.rst:4]: @maddenp-noaa This looks fine to me.

File: docs/sections/user_guide/cli/drivers/cdeps/help.out
Comment (uturuncoglu) [inline docs/sections/user_guide/cli/drivers/cdeps/help.out:1]: @maddenp-noaa As you can see from here. The current implementation is only able to create DATM and DOCN namelist files. Actually, CDEPS has also data wave, data land etc. but data ocean and atmosphere are most used ones. So, it would be nice to add some notes to the CDEPS part of the documentation. Maybe we could extend the implementation later to include others if somebody needs it. It is just a matter of updating the scheme file and make minor mods in the CDEPS driver.

File: docs/sections/user_guide/yaml/components/cdeps.rst
Comment (uturuncoglu) [reply docs/sections/user_guide/yaml/components/cdeps.rst]: @maddenp-noaa The following section of CDEPs documentation includes information about the namelist options: https://escomp.github.io/CDEPS/versions/master/html/streams.html

File: docs/shared/cdeps.yaml
Comment (uturuncoglu) [reply docs/shared/cdeps.yaml:2]: @maddenp-noaa That is great. Thanks.

File: src/uwtools/cli.py
Comment (uturuncoglu) [reply src/uwtools/cli.py:855]: @maddenp-noaa It looks great. I think this is one step closer to what we achieve.

File: src/uwtools/resources/jsonschema/cdeps.jsonschema
Comment (uturuncoglu) [reply src/uwtools/resources/jsonschema/cdeps.jsonschema:3]: @maddenp-noaa Thanks. Yes, some part of those books are similar but they could also have minor differences.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (uturuncoglu) [reply src/uwtools/tests/drivers/test_cdeps.py:3]: @maddenp-noaa That is great. I was not aware of having test for drivers. Thanks for adding it.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (uturuncoglu) [reply src/uwtools/tests/drivers/test_cdeps.py:94]: @maddenp-noaa Okay. I was wondering about the Jinja2 template. So, users could just generate it by running tests.

File: src/uwtools/tests/test_schemas.py
Comment (uturuncoglu) [reply src/uwtools/tests/test_schemas.py:15]: @maddenp-noaa Some of them are list such as datamode can be only certain values since CDEPS supports set of data modes for each data model. It is hard to put constraint for others such as model_maskfile and model_meshfile. These are just need to point a ESMF mesh file that represent the data file (or domain). I am not sure what is the best approach in here.

**Review by uturuncoglu — APPROVED** [review]
@maddenp-noaa Thanks for your help. This looks great. Lots of missing pieces are glued together. I put some minor comments but it is up to you to change anything based on it. So, I am approving this but let me know when you merge this to my branch.

File: src/uwtools/tests/drivers/test_cdeps.py
Comment (maddenp-cu) [reply src/uwtools/tests/drivers/test_cdeps.py:94]: For other drivers where we must currently rely on a Jinja2 template, we're expecting that the Apps that make use of `uwtools` will provide the appropriate template file(s) in their repo. Then the UW YAML for the component (CDEPS in this case) can point to that template file in the App's git clone (for example) as the source.

File: src/uwtools/tests/test_schemas.py
Comment (maddenp-cu) [reply src/uwtools/tests/test_schemas.py:15]: We can move forward with what's here and of course adjust it later if we develop better ideas.

**Comment by maddenp-cu** [comment]
Hi @uturuncoglu -- I made a couple of small changes in the documentation links, but I think this is ready to merge. You'll have to do the merge in GitHub, as I do not have write access to your repo. Once you do, I can encourage the rest of my team to review your main PR targeting the `uwtools` repo.

**Comment by uturuncoglu** [comment]
@maddenp-noaa I have just merged. Thanks for all your help.

**Pull request merged**
